src/Build.UnitTests/Graph/ProjectGraph_Tests.cs

```diff
@@ -600,8 +628,232 @@ public void GetTargetListsNullEntryTargets()
+
+                var dot = projectGraph.ToDot();
+
+                projectGraph.ProjectNodes.Count.ShouldBe(2);
+
+                IReadOnlyDictionary<ProjectGraphNode, ImmutableList<string>> targetLists = projectGraph.GetTargetLists(new List<string>{"A"});
+                targetLists.Count.ShouldBe(projectGraph.ProjectNodes.Count);
+
+                targetLists[GetFirstNodeWithProjectNumber(projectGraph, 1)].ShouldBe(new[] { "A" });
+                targetLists[GetFirstNodeWithProjectNumber(projectGraph, 2)].ShouldBe(NonOuterBuildTargets);
+            }
+        }
+
+        [Fact]
+        public void GetTargetListsFiltersTargetsForOuterAndInnerBuilds()
+        {
+            using (var env = TestEnvironment.Create())
+            {
+                var root1 = CreateProjectFile(env, 1, new[] {2}, null, null, ProjectReferenceTargetsWithMultitargeting).Path;
+                CreateProjectFile(env, 2, null, null, null, OuterBuildSpecificationWithProjectReferenceTargets);
+                
+                var projectGraph = new ProjectGraph(root1);
+
+                var dot = projectGraph.ToDot();
+
+                projectGraph.ProjectNodes.Count.ShouldBe(4);
+
+                IReadOnlyDictionary<ProjectGraphNode, ImmutableList<string>> targetLists = projectGraph.GetTargetLists(new List<string>{"A"});
+
+                targetLists.Count.ShouldBe(projectGraph.ProjectNodes.Count);
+                var root = GetFirstNodeWithProjectNumber(projectGraph, 1);
+
+                var outerBuild = GetNodesWithProjectNumber(projectGraph, 2).First(IsOuterBuild);
+                var innerBuilds = GetNodesWithProjectNumber(projectGraph, 2).Where(IsInnerBuild).ToArray();
+
+                targetLists[root].ShouldBe(new[] { "A" });
+                targetLists[outerBuild].ShouldBe(OuterBuildTargets);
+
+                foreach (var innerBuild in innerBuilds)
+                {
+                    targetLists[innerBuild].ShouldBe(NonOuterBuildTargets);
+                }
+            }
+        }
+
+        [Fact]
+        public void GetTargetListsForComplexMultitargetingGraph()
+        {
+            using (var env = TestEnvironment.Create())
+            {
+                var root1 = CreateProjectFile(
+                    env,
+                    1,
+                    null,
+                    null,
+                    null,
+                    OuterBuildSpecificationWithProjectReferenceTargets +
+                    $@"<ItemGroup>
+                            <ProjectReference Include=`3.proj` Condition=`'$({InnerBuildProperty})'=='a'`/>
+
+                            <ProjectReference Include=`4.proj` Condition=`'$({InnerBuildProperty})'=='b'`/>
+                            <ProjectReference Include=`5.proj` Condition=`'$({InnerBuildProperty})'=='b'`/>
+                            <ProjectReference Include=`6.proj` Condition=`'$({InnerBuildProperty})'=='b'` Properties=`{InnerBuildProperty}=a`/>
```

Comment:
I'm confused on this one. 

![image](https://user-images.githubusercontent.com/3347530/54761733-a5dc6800-4bc0-11e9-855c-dcbbafe3b930.png)

Why do we get the speculative `[p1 ib=b] -> [p6 outer]` and `[p6 ib=a]` edges? My read on https://github.com/Microsoft/msbuild/blob/56e10566f8491faab2421221f9b68b109b06a014/documentation/specs/static-graph.md#multitargeting was that only references to outer builds (subject to TF negotiation) would get the speculative links.

Reply:
(Also, while I agree with Andy that `ToDot()` should stay internal--it's awesome :))

Reply:
Wow, good catch! It appears to be a bug. Maybe. `[p1 ib=b]` is directly referencing the inner build `[p6 ib=a]` by setting `Properties="InnerBuild=a"` metadata on `ProjectReference`. But at the same time, inner builds set the inner build property as an `UndefineProperty`, to avoid propagating their TF to the references. This undefine removes the effect of `Properties="InnerBuild=a"` and leads to P6 being transformed into an outer build.

I'll go debug how vanilla msbuild handles this situation, and adjust to that behaviour.

Reply:
Mkay, this is how the sdk works (inner build A(net472) references B(net462) via `Properties="TargetFramework=net462"` metadata):
- A gets to call B(no global props), the outer build, with `GetTargetFrameworks` (GTF) and B(net462) with Build, GetNativeManifest, etc
  - Calling B(no global props) with GTF is wasted work, as the information won't get used since the Properties metadata overwrites any TF negotiation
  - A(net472) calls GTF on B outer build because [TargetFramework gets explicitly removed for calls to GTF](https://github.com/Microsoft/msbuild/blob/a972ec96c3920705e4e8d03d7ac8b6c3328450bd/src/Tasks/Microsoft.Common.CurrentVersion.targets#L1642)
- The scenario ends up working in reality because the [SDK does not remove `TargetFramework`](https://github.com/Microsoft/msbuild/blob/a972ec96c3920705e4e8d03d7ac8b6c3328450bd/src/Tasks/Microsoft.Common.CurrentVersion.targets#L1686-L1688) for the inner targets (Build, GetNativeManifest etc) when Nuget finds a `NearestTargetFramework` and the reference has multiple target frameworks. The static graph, on the other hand, always removes the `TargetFramework` on an inner build's references, which leads to the B(net462) getting converted to an outer build instead.

So, in this case, the static graph accidentally overbuilds, which I guess is better than underbuilding, which would cause cache miss build breaks at runtime. Since having inner builds manually reference other inner builds seems like an edge case, I propose fixing it in a separate PR. Let me know if you think otherwise.

Reply:
Separate PR sounds fine--can you file an issue please?

Reply:
https://github.com/Microsoft/msbuild/issues/4262